_posts/2025-07-08-quarkus-and-a2a-java-sdk.adoc

```diff
@@ -0,0 +1,283 @@
+
+=== Important Concepts
+
+The A2A protocol involves a few important concepts:
+
+* *User* - This is the end user who has a request that will require the help of one or more agents.
+* *A2A Client* - This is the client that will send requests on the user's behalf to an A2A server agent.
+* *A2A Server* - This is the server agent that will receive and respond to requests from an A2A client agent. An A2A server agent exposes an HTTP endpoint that implements the A2A protocol.
+
+A2A client agents and A2A server agents can be implemented using different languages and frameworks. They just need to be able to speak with each other using the A2A protocol. Communication happens using JSON-RPC 2.0 over HTTP(S). A2A SDKs written for different programming languages make this interoperability possible.
+
+The https://github.com/orgs/a2aproject/repositories[A2A project] aims to provide SDKs for various languages. Using the https://github.com/a2aproject/a2a-python[A2A Python SDK] and our https://github.com/a2aproject/a2a-java[A2A Java SDK], for example, it's possible for an A2A client agent written in Python to communicate with an A2A server agent written in Java and vice versa.
+
+== From a Quarkus LangChain4j AI Service to an A2A Server Agent
+
+Let's say we have a Quarkus LangChain4j AI service that can respond to user queries about the weather by making use of a weather MCP server:
+
+[source,java]
+----
+@RegisterAiService
+@ApplicationScoped
+public interface WeatherAgent {
+
+    @SystemMessage("""
+            You are a specialized weather forecast assistant. Your primary function is to
+            utilize the provided tools to retrieve and relay weather information in response
+            to user queries. You must rely exclusively on these tools for data and refrain
+            from inventing information. Ensure that all responses include the detailed output
+            from the tools used and are formatted in Markdown.
+            """
+    )
+    @McpToolBox("weather") // <-- The weather MCP server that will be used
+    String chat(@UserMessage String question);
+}
+----
+
+To turn this weather agent into an A2A server agent, there are a few simple steps we need to follow:
+
+=== Add an A2A Java SDK Server Dependency
+
+[NOTE]
+====
+The `io.github.a2asdk` `groupId` is temporary and will likely change for future releases. Keep an eye on the `a2a-java` https://github.com/a2aproject/a2a-java/blob/main/README.md[README] for up-to-date documentation.
+====
+
+[source,xml]
+----
+<dependency>
+    <groupId>io.github.a2asdk</groupId>
+    <artifactId>a2a-java-sdk-server-quarkus</artifactId> <1>
+</dependency>
+----
+<1> `a2a-java-sdk-server-quarkus` provides access to the core classes that make up the A2A specification and provides the HTTP endpoint that implements the A2A protocol. This dependency makes use of Quarkus Reactive Routes. To make use of Jakarta REST, the `a2a-java-sdk-server-jakarta` dependency can be used instead.
```

Comment:
is this really using Quarkus Reactive Routes and not just jaxrs with reactive endpoints?

we dont really put jakarta in names historically thus just trying to grok the diff?

@geoand @cescoffier ^

Reply:
`a2a-java-sdk-server-quarkus` has a dependency on `quarkus-reactive-routes` and makes use of the `@Route` annotation (see https://github.com/a2aproject/a2a-java/blob/main/sdk-quarkus/src/main/java/io/a2a/server/apps/quarkus/A2AServerRoutes.java#L77).

`a2a-java-sdk-server-jakarta` makes use of jaxrs with `@POST` annotations (e.g., see https://github.com/a2aproject/a2a-java/blob/main/sdk-jakarta/src/main/java/io/a2a/server/apps/jakarta/A2AServerResource.java).

@kabir @maeste ^

Reply:
> is this really using Quarkus Reactive Routes and not just jaxrs with reactive endpoints?

It is.

Reply:
+1

Reply:
Just to clarify, sdk-quarkus and sdk-jakarta are mutually exclusive.

sdk-quarkus should be used if running in Quarkus, and that uses Reactive Routes.

sdk-jakarta uses Jakarta RESTful web services, and is intended for running in WildFly, Tomcat etc. We currently test that with WildFly. Originally we tested that with Quarkus with 'legacy' quarkus-resteasy, but found when deploying in WildFly that we had to make some [minor changes](https://github.com/a2aproject/a2a-java/pull/144) to the CDI beans for them to be valid in WildFly. Thus we thought it best to test that part in WildFly